tests/nativeapp/test_artifacts.py

```diff
@@ -163,3 +168,51 @@ def test_too_many_files(project_definition_files):
                 ArtifactMapping("app/streamlit/*.py", "somehow_combined_streamlits.py")
             ],
         )
+
+
+@pytest.mark.parametrize(
+    "project_path,expected_path",
+    [
+        [
+            "file",
+            "deploy/file",
+        ],
+        [
+            "dir",
+            "deploy/dir",
+        ],
+        [
+            "dir/nested_file1",
+            "deploy/dir/nested_file1",
+        ],
+        [
+            "dir/nested_dir/nested_file2",
+            "deploy/dir/nested_dir/nested_file2",
+        ],
+        [
+            "dir/nested_dir",
+            "deploy/dir/nested_dir",
+        ],
+    ],
+)
+def test_project_path_to_deploy_path(
+    temp_dir,
+    project_path,
+    expected_path,
+):
+    # Source files
+    touch("file")
+    touch("dir/nested_file1")
+    touch("dir/nested_dir/nested_file2")
+    # Build
+    os.mkdir("deploy")
+    os.symlink("file", "deploy/file")
+    os.symlink(Path("dir").resolve(), Path("deploy/dir"))
+
+    files_mapping = {
```

Comment:
These tests don't cover the full contract. I'd add test cases where the directory was renamed too, and a case where no mapping is found.

Reply:
When the user specifies a file that exists in project root but isn't mapped by any artifact pattern, `FileNotFoundError` will be raised. Added a test for it.

Not sure I understand the renamed directory test -- can you explain?

Reply:
I mean foo in src goes to bar in dest. To make sure we compose paths correctly. Currently your tests assume src & dest have the same names.

Reply:
Got it. Changed the file/dir names in these tests to demonstrate they don't have to match.